### ISSUE TITLE
Operation Timed out on ROSE2

### ISSUE THREAD

**Issue by GryderArt** [BradnerLab/pipeline#62]
An unusual error is clogging ROSE2 during bam mapping.  The log file reads:

MAPPING TO THE FOLLOWING BAMS:
['/data/khanlab/projects/ChIP_seq/DATA/H3K27ac_024_C_HLFMLBGX3.bam', '/data/khanlab/projects/ChIP_seq/DATA/input_026_C_H7FGKBGX5.bam']
OPERATION TIMED OUT. FILE /data/khanlab/projects/ChIP_seq/DATA/ROSE_out_12500/mappedGFF/Sample_KG1a_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_STITCHED_TSS_DISTAL_H3K27ac_024_C_HLFMLBGX3.bam_MAPPED/matrix.txt NOT FOUND

Has anyone seen this error before, and solved it?
I'm running ROSE2 on a cluster node with --mem=121g --cpus-per-task=4

**Comment by GryderArt** [comment]
moved to Charles Lin site where they see the same issue
https://github.com/linlabbcm/rose2/issues/6

**Comment by charlesylin** [comment]
Berkley,

Do you see any output at all in this folder?

/data/khanlab/projects/ChIP_seq/DATA/ROSE_out_12500/
mappedGFF/Sample_KG1a_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_
STITCHED_TSS_DISTAL_H3K27ac_024_C_HLFMLBGX3.bam_MAPPED/

Charles Y. Lin
Assistant Professor
Department of Molecular and Human Genetics
Dan L. Duncan Cancer Center
Baylor College of Medicine

**Comment by GryderArt** [comment]
No - normally I see files in that folder, but currently they are being left blank. in fact, there are no folders at all underneath the mappedGFF folder.

**Comment by charlesylin** [comment]
So that's the step that bamliquidator is running. Do you see anything at
all in standard output showing either an error for bamliquidator or a
"liquidating..." style output?

My suspicion is that ROSE2 can't find/call bamliquidator correctly on the
node.

-Charles

Charles Y. Lin
Assistant Professor
Department of Molecular and Human Genetics
Dan L. Duncan Cancer Center
Baylor College of Medicine

**Comment by GryderArt** [comment]
there is no liquidating style output:
![image](https://user-images.githubusercontent.com/11543307/38205154-8644b6e0-3673-11e8-88a0-ab27d32f5da5.png)

but I am running this on a cluster node as a batch job, and thus it is possible that I'm not able to see everything I would if I ran it on an interactive node.  Should I try that to see what other errors are thrown?

**Comment by GryderArt** [comment]
ok so here are some more lines of error following the first "Operation Timed Out" error:
MAPPING TO THE FOLLOWING BAMS:
['/data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/Sample_H3K27ac_024_C_HLFMLBGX3.bam', '/data/khanlab/projects/ChIP_seq/DATA/Sample_D4_input_026_C_H7FGKBGX5/Sample_D4_input_026_C_H7FGKBGX5.bam']
OPERATION TIMED OUT. FILE /data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/MACS_Out_p_1e-14/ROSE_out_12500/mappedGFF/Sample_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_STITCHED_TSS_DISTAL_Sample_H3K27ac_024_C_HLFMLBGX3.bam_MAPPED/matrix.txt NOT FOUND
bamliquidator_batch --sense . -e 200 --match_bamToGFF -r /data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/MACS_Out_p_1e-14/ROSE_out_12500/gff/Sample_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_STITCHED_TSS_DISTAL.gff -o /data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/MACS_Out_p_1e-14/ROSE_out_12500/mappedGFF/Sample_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_STITCHED_TSS_DISTAL_Sample_H3K27ac_024_C_HLFMLBGX3.bam_MAPPED /data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/Sample_H3K27ac_024_C_HLFMLBGX3.bam
OPERATION TIMED OUT. FILE /data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/MACS_Out_p_1e-14/ROSE_out_12500/mappedGFF/Sample_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_STITCHED_TSS_DISTAL_Sample_H3K27ac_024_C_HLFMLBGX3.bam_MAPPED/matrix.txt NOT FOUND
ERROR: FAILED TO MAP /data/khanlab/projects/ChIP_seq/DATA/Sample_H3K27ac_024_C_HLFMLBGX3/MACS_Out_p_1e-14/ROSE_out_12500/gff/Sample_H3K27ac_024_C_HLFMLBGX3_peaks_0KB_STITCHED_TSS_DISTAL.gff FROM BAM: Sample_H3K27ac_024_C_HLFMLBGX3.bam

**Comment by charlesylin** [comment]
Yeah. It definitely cannot find bamliquidator. Not sure why this is breaking all of a sudden when it was working previously. 

Can you run bamliquidator on its own and see if it works?

-Charles

**Comment by GryderArt** [comment]
We found the 5 minute max time hardcoded into ROSE here:
if utils.checkOutput(mappedOut1File,0.2,5):
                print("SUCCESSFULLY MAPPED TO %s FROM BAM: %s" % (stitchedGFFFile, bamFileName))
            else:
                print("ERROR: FAILED TO MAP %s FROM BAM: %s" % (stitchedGFFFile, bamFileName))

And tried to increase the max run time from 5 minutes to 30 minutes – turns out that didn’t fix it, and since bamliquidator can run at more than 11 million reads per second, and our BAM files don’t have more than 40 million reads each, it is not a time out of running bamliquidator.

**Comment by GryderArt** [comment]
also, bamliqidator is running just fine from the command line on this sample.  Just isn't working from ROSE2.

**Comment by GryderArt** [comment]
so we got a “segmentation fault” when running bamliquidator_batch.py on an interactive node; what does that mean?

**Comment by charlesylin** [comment]
nothing good. it's possible bamliquidator was compiled on one node and is
not compatible on the other, but this is now starting to get outside of my
realm of expertise.  am cc'ing john dimatteo who developed bamliquidator.

John is there some sort of output that would be most informative to debug a
seg fault or should they just try to re-compile on the node?

Unfortunately this is getting into territory where it's highly dependent on
OS and other base level configuration.

Charles Y. Lin
Assistant Professor
Department of Molecular and Human Genetics
Dan L. Duncan Cancer Center
Baylor College of Medicine

**Comment by GryderArt** [comment]
So, while I don't understand it entirely, we got it working by correcting the version control of bamliquidator.
during the failures, it was calling up in this manner:
[+] Loading hdf5 1.8.15
[+] Loading bamliquidator version 1.3...

now, hdf5 doesn't even appear in the loading modules list.  we are using something called "singularity", which loads a uniform version:
[+] Loading bamliquidator 1.3.4 on cn1654
[+] Loading singularity 2.4.5 on cn1654

The explanation I got from the folks managing our clusters:

> Bamliquidator v.1.3.4 (the default now) is in a singularity container. By loading the module it gives you bamliquidator and bamliquidator_batch as executables/wrappers, which calls bamliquidator inside the container. Singularity is a type of container (just like docker). In the case of bamliquidator, it allows bamliquator to be installed using ubuntu OS (using apt-get) and used/run in the HPC centos OS. More details on: https://hpc.nih.gov/apps/singularity.html
